File: python_client/net/connection_handler.py
"""
Connection handler: performs the full handshake then routes encrypted messages.
"""
from net.framing import decode_frame, FramingError
from protocol.serializer import json_loads_bytes


def handle_connection(conn, addr, identity=None, ctx=None):
    """
    Handle an incoming TCP connection.

    If an identity and context are provided, performs the full server-side
    handshake then routes messages through the MessageRouter.
    Falls back to plain message printing for compatibility testing.

    Args:
        conn: The accepted socket connection
        addr: The remote address tuple
        identity: The local Identity object (for handshake)
        ctx: The runtime context dict
    """
    print(f"[NET] Connection from {addr}")

    if identity is None:
        # Plain mode: just read and print (for testing/compat)
        stream = conn.makefile("rb")
        try:
            while True:
                try:
                    payload = decode_frame(stream)
                except FramingError:
                    break
                msg = json_loads_bytes(payload)
                print(f"[NET] Received: {msg}")
        finally:
            try:
                stream.close()
            except Exception:
                pass
            conn.close()
        return

    try:
        from net.handshake_server import execute_server_handshake
        session, client_hello = execute_server_handshake(conn, identity)
        peer_id = client_hello.get("peer_id", "unknown")
        peer_name = client_hello.get("peer_name", "unknown")
        print(f"[AUTH] Handshake complete.")
        print(f"  Peer name       : {peer_name}")
        print(f"  Peer fingerprint: {peer_id}")
        print(f"  Session ID      : {session.session_id}")
        print()
        print(f"[AUTH] Mutual authentication successful.")
        print(f"  - We verified {peer_name}'s RSA-PSS signature.")
        print(f"  - {peer_name} verified our RSA-PSS signature.")
        print(f"  - Session keys derived via ephemeral Diffie-Hellman (new keys, never stored).")
        print(f"  - All further communication is AES-256-GCM encrypted.")

        # Save verified contact
        if ctx is not None and "contacts_store" in ctx:
            contacts = ctx["contacts_store"].load()
            contacts[peer_id] = {
                "peer_name": peer_name,
                "rsa_public_key_der_b64": client_hello.get("rsa_public_key_der_b64"),
            }
            ctx["contacts_store"].save(contacts)

        # Route all subsequent messages through the MessageRouter
        from net.router import MessageRouter
        router = MessageRouter(ctx, conn, session, peer_name, peer_id)
        router.run()

    except Exception as e:
        print(f"[ERROR] Connection from {addr} failed: {e}")
    finally:
        try:
            conn.close()
        except Exception:
            pass

File: java_client/src/main/java/com/cisc468share/net/ConnectionHandler.java
package com.cisc468share.net;

import com.cisc468share.crypto.HandshakeManager;
import com.cisc468share.files.ShareManager;
import com.cisc468share.crypto.SecureSession;
import com.cisc468share.router.MessageRouter;

import java.net.Socket;

/**
 * Handles an incoming TCP connection:
 * 1. Performs the server-side cryptographic handshake (authenticates the peer).
 * 2. Saves the peer's verified identity to contacts.
 * 3. Delegates all subsequent messages to the MessageRouter.
 */
public class ConnectionHandler {

    private final HandshakeManager handshakeManager;

    /** Construct with identity information for the handshake. */
    private final ShareManager shareManager;

    /** Construct with identity information for the handshake. */
    public ConnectionHandler(HandshakeManager handshakeManager, ShareManager shareManager) {
        this.handshakeManager = handshakeManager;
        this.shareManager = shareManager;
    }

    /** Construct without identity — plain echo mode for testing. */
    public ConnectionHandler() {
        this.handshakeManager = null;
        this.shareManager = null;
    }

    public void handle(Socket socket) {
        System.out.println("[NET] Connection from " + socket.getRemoteSocketAddress());

        if (handshakeManager == null) {
            // Plain mode (compatibility / testing)
            handlePlain(socket);
            return;
        }

        try {
            SecureSession session = handshakeManager.executeServerHandshake(socket);
            System.out.println("[NET] Handshake complete — session: " + session.sessionId);

            MessageRouter router = new MessageRouter(socket, session, "peer", "unknown", shareManager);
            router.run();

        } catch (Exception e) {
            System.out.println("[ERROR] Connection failed: " + e.getMessage());
        } finally {
            try { socket.close(); } catch (Exception ignored) {}
        }
    }

    private void handlePlain(Socket socket) {
        try {
            java.io.InputStream in = socket.getInputStream();
            while (true) {
                byte[] payload = Framing.decodeFrame(in);
                System.out.println("[NET] Received " + payload.length + " bytes");
            }
        } catch (Exception e) {
            System.out.println("[NET] Connection closed");
        } finally {
            try { socket.close(); } catch (Exception ignored) {}
        }
    }
}
